### DISCUSSION TITLE
function()

### DISCUSSION THREAD

**Discussion by gkim132** [jspsych/jsPsych#863] [Q&A] [answered]
I'm trying to assign the correct_response, either 'f' or 'j', based on the file name. If a file name has 'J' at the last 5th, then I want to assign the correct_response to 'j', otherwise 'f' 

Any help will be greatly appreciated!
Thank you!

`
var images_1 = ["t_img_1/1.gif", "t_img/2.gif","t_img/3_J.gif","t_img/4.gif","t_img/5.gif","t_img/6_J.gif","t_img/7_J.gif","t_img/8_J.gif","t_img/9.gif","t_img/10.gif","t_img/11_J.gif","t_img/12.gif"];
`
`    
var images_2 = ["t_img_2/1.gif", "t_img_2/2.gif","t_img_2/3.gif","t_img_2/4.gif","t_img_2/5.gif","t_img_2/6.gif","t_img_2/7.gif","t_img_2/8.gif","t_img_2/9.gif","t_img_2/10.gif","t_img_2/11.gif","t_img_2/12.gif"];
 `

    var trials = [];
    for (var i = 0; i < images_1.length; i++) {
      trials.push({
        stimulus: `<div>
                <div style="display:inline-block"><img
                style="width: 30vw;"
                src=  ${ images_1[i] } ></div>
                <div style="display:inline-block"><img 
                style="width: 30vw;"
                src= ${ images_1[i] }></div>
                </div>`,
        //data: { test_part: 'test', correct_response: 'j'}  //This will assgin the correct_response of ALL trials to 'j' , that is I don't want
        data: function(){                                    //This doesn't work
          var myString = images_1[i];
          var lastChar = myString[myString.length -5];
          if ( lastCar == 'J'){
            return {test_part: 'test', correct_response: 'j'};
          }
          else{
            return {test_part: 'test', correct_response: 'f'};
          }
          
        }
      });
    }
    var test = {
      type: 'html-keyboard-response',
      choices: ['f', 'j'],
      on_finish: function(data){
        data.correct = data.key_press == jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.correct_response);
      },

      timeline: trials
    }

    var after_block = {
      type: 'html-keyboard-response',
      stimulus: '<p>The next node</p>',
      //is_html: true
    }
 
    
    jsPsych.init({
      timeline: [ test, after_block],
      //timeline: timeline
      on_finish: function() {
        jsPsych.data.displayData();
      }
    });`

**Comment by jodeleeuw** [comment]
Hi @gkim132 

What happens when you run the code as it is right now?

**Reply by gkim132** [reply]
i have this error:
Uncaught TypeError: Cannot read property 'length' of undefined
    at data (t.html:35)
    at evaluateFunctionParameters (jspsych.js:902)
    at doTrial (jspsych.js:834)
    at startExperiment (jspsych.js:784)
    at jspsych.js:2569
    at Object.jsPsych.pluginAPI.module.preloadVideo (jspsych.js:2449)
    at jspsych.js:2568
    at Object.jsPsych.pluginAPI.module.preloadAudioFiles (jspsych.js:2295)
    at jspsych.js:2567
    at Object.jsPsych.pluginAPI.module.preloadImages (jspsych.js:2400)

**Comment by jodeleeuw** [comment, marked as answer]
The problem has to do with when functions execute in the code.

Currently you are constructing your trials in a for-loop and using a function as the value to the `data` parameter. When you use a function as a parameter, the function runs during the experiment. This means that the variable `i` in your function is not updating as you expect in the loop. The loop is actually done long-before the function ever runs, and I'm guessing that `i` is equal to `images_1.length` when the function runs because that would be the last value of `i` when the loop is ended. Or maybe there's a scope issue here and `i` is undefined. Either way, the solution is the same!

You'll want to run your data parameter function during the construction of the timeline. One way to do that would be to move it to the top of the for loop and then use the variable it generates in the data parameter:

```js
for (var i = 0; i < images_1.length; i++) {
  var myString = images_1[i];
  var lastChar = myString[myString.length -5];
  if ( lastCar == 'J'){
    var trial_data =  {test_part: 'test', correct_response: 'j'};
  }
  else{
    var trial_data = {test_part: 'test', correct_response: 'f'};
  }
  trials.push({
    stimulus: `<div>
            <div style="display:inline-block"><img
            style="width: 30vw;"
            src=  ${ images_1[i] } ></div>
            <div style="display:inline-block"><img 
            style="width: 30vw;"
            src= ${ images_1[i] }></div>
            </div>`,
    data: trial_data
  });
}
```

**Reply by gkim132** [reply]
It works now! Thank you so much @jodeleeuw
